File: sdk/dotnet/Shojiku/Configuration.cs
// Process-wide configuration, and the entry points that reach it.
//
// The ecosystem idiom (a Configure call in application start-up) OVER the
// constructor, never a third precedence layer: what Configure sets stands
// exactly where an explicit constructor argument stands against the environment.
// So the order is
//
//     explicit argument > Shojiku.Configuration.Configure > SHOJIKU_*
//
// for the template root and the pack directories, and the deliberate reverse for
// the engine library — SHOJIKU_LIBRARY still wins over both, because where the
// engine lives is a deployment decision.
//
// `Strict` is the one exception, and it is the only place Configure beats a call
// site. Strictness is a restriction rather than a default: an operator who
// declared a lockdown must not have it lifted by application code passing
// strict: false. Every SDK mirrors that asymmetry.
//
// The rule the other six mirror: an ecosystem-standard configuration idiom feeds
// the same constructor and never adds a precedence level of its own.

namespace Shojiku;

/// <summary>Process-wide defaults for every client built after it is set.</summary>
public sealed class Config
{
    /// <summary>The directory template names resolve against.</summary>
    public string? Templates { get; set; }

    /// <summary>Directories holding font packs.</summary>
    public IReadOnlyList<string>? FontDirs { get; set; }

    /// <summary>Directories holding locale packs.</summary>
    public IReadOnlyList<string>? LocaleDirs { get; set; }

    /// <summary>The BCP 47 locale a render defaults to.</summary>
    public string? Lang { get; set; }

    /// <summary>An explicit path to the engine library. <c>SHOJIKU_LIBRARY</c> still beats it.</summary>
    public string? Library { get; set; }

    /// <summary>Where the host-side log channel writes, when an application wants one.</summary>
    public IShojikuLogger? Logger { get; set; }

    /// <summary>Whether clients refuse the bytes entrance and unregistered providers.</summary>
    public bool Strict { get; set; }

    /// <summary>The signing providers a strict client may be asked for by name.</summary>
    public IReadOnlyDictionary<string, object>? Providers { get; set; }

    /// <summary>Whether <c>SHOJIKU_*</c> lookups happen at all.</summary>
    public bool Env { get; set; } = true;

    /// <summary>
    /// A copy with <paramref name="overrides"/> applied — one client's resolution step.
    /// </summary>
    /// <remarks>
    /// A null override means "not given", so an explicit constructor argument
    /// beats a configured default and an absent one inherits it.
    /// <see cref="Strict"/> is the exception documented above: it is OR-ed
    /// rather than overridden.
    /// <para>
    /// <see cref="Providers"/> REPLACES rather than merges. A client that
    /// declares its own registry is stating the whole set it may sign with, and
    /// quietly adding globally-registered keys to that set would defeat the
    /// point.
    /// </para>
    /// </remarks>
    internal Config Merge(ClientOptions overrides) => new()
    {
        Templates = overrides.Templates ?? Templates,
        FontDirs = overrides.FontDirs ?? FontDirs,
        LocaleDirs = overrides.LocaleDirs ?? LocaleDirs,
        Lang = overrides.Lang ?? Lang,
        Library = overrides.Library ?? Library,
        Logger = overrides.Logger ?? Logger,
        Providers = overrides.Providers ?? Providers,
        Env = overrides.Env ?? Env,
        Strict = Strict || (overrides.Strict ?? false),
    };
}

/// <summary>What one client was constructed with, before the defaults are merged in.</summary>
internal sealed record ClientOptions(
    string? Templates = null,
    IReadOnlyList<string>? FontDirs = null,
    IReadOnlyList<string>? LocaleDirs = null,
    string? Lang = null,
    string? Library = null,
    IShojikuLogger? Logger = null,
    bool? Strict = null,
    IReadOnlyDictionary<string, object>? Providers = null,
    bool? Env = null);

/// <summary>The process-wide defaults, and the entry points that set them.</summary>
public static class Configuration
{
    private static Config current = new();

    /// <summary>The process-wide defaults, read by every client at construction.</summary>
    public static Config Current => current;

    /// <summary>
    /// Set process-wide defaults.
    /// </summary>
    /// <example>
    /// <code>
    /// Shojiku.Configuration.Configure(config =>
    /// {
    ///     config.Templates = "App/Templates";
    ///     config.Lang = "ja-JP";
    /// });
    /// </code>
    /// </example>
    /// <param name="configure">Mutates the process-wide defaults.</param>
    /// <returns>The configured defaults.</returns>
    public static Config Configure(Action<Config> configure)
    {
        configure(current);
        return current;
    }

    /// <summary>
    /// Drop every configured default.
    /// </summary>
    /// <remarks>
    /// Public because a global that cannot be reset makes every test suite
    /// invent its own teardown — and get it wrong in a randomly-ordered run.
    /// Applications call it at most once, if at all.
    /// </remarks>
    public static void Reset() => current = new Config();
}

File: sdk/php/src/Configuration.php
<?php

declare(strict_types=1);

namespace Shojiku;

use Shojiku\Exception\UsageException;

/**
 * Process-wide defaults for every {@see Client} built after they are set.
 *
 * The ecosystem idiom (an options array applied once during bootstrap) OVER
 * the frozen constructor, never a third precedence layer: what
 * {@see self::configure()} sets stands exactly where an explicit constructor
 * argument stands against the environment. So the order is
 *
 *     explicit argument > Configuration::configure() > SHOJIKU_*
 *
 * for the template root and the pack directories, and the deliberate reverse
 * for the engine binary — `SHOJIKU_BIN` still wins over both, because where
 * the engine lives is a deployment decision that has to be able to win over
 * application code.
 *
 * **`strict` is the one exception, and it is the only place `configure()`
 * beats a call site.** Strictness is a restriction rather than a default: an
 * operator who declared a lockdown must not have it lifted by application
 * code passing `strict: false`. Every SDK mirrors that asymmetry.
 *
 * An ARRAY rather than a mutable object handed to a closure, because that is
 * what lets a misspelled key be a named error: assigning an undeclared
 * property is only a deprecation notice in PHP, so the object form would
 * silently ignore `config->template = …` and render nothing.
 */
final class Configuration
{
    /**
     * Every setting a client can take, which is also what
     * {@see self::configure()} accepts — so a misspelled key is a named error
     * rather than a silently ignored one.
     */
    public const SETTINGS = [
        'templates', 'fontDirs', 'localeDirs', 'lang', 'binary', 'logger', 'strict', 'providers', 'env',
    ];

    private static ?self $global = null;

    /**
     * @param list<string>|null $fontDirs
     * @param list<string>|null $localeDirs
     * @param array<string, SigningProvider>|null $providers
     */
    public function __construct(
        public readonly ?string $templates = null,
        public readonly ?array $fontDirs = null,
        public readonly ?array $localeDirs = null,
        public readonly ?string $lang = null,
        public readonly ?string $binary = null,
        public readonly ?object $logger = null,
        public readonly bool $strict = false,
        public readonly ?array $providers = null,
        public readonly bool $env = true,
    ) {
    }

    /**
     * The process-wide defaults, read by every {@see Client} at construction.
     */
    public static function global(): self
    {
        return self::$global ??= new self();
    }

    /**
     * Sets the process-wide defaults.
     *
     * ```php
     * Shojiku\Configuration::configure(['templates' => 'app/templates', 'lang' => 'ja-JP']);
     * ```
     *
     * @param array<string, mixed> $options
     *
     * @throws UsageException on an unknown setting name
     */
    public static function configure(array $options): self
    {
        return self::$global = self::global()->merge($options);
    }

    /**
     * Drops every configured default.
     *
     * Public because a global that cannot be reset makes every test suite
     * invent its own teardown — and get it wrong. Applications call it at
     * most once, if at all.
     */
    public static function reset(): void
    {
        self::$global = null;
    }

    /**
     * A copy with `$overrides` applied — one client's resolution step.
     *
     * A null override means "not given", so an explicit constructor argument
     * beats a configured default and an absent one inherits it. `strict` is
     * the exception documented above: it is OR-ed rather than overridden.
     *
     * `providers` REPLACES rather than merges. A client that declares its own
     * registry is stating the whole set it may sign with, and quietly adding
     * globally-registered keys to that set would defeat the point.
     *
     * @param array<string, mixed> $overrides
     *
     * @throws UsageException on an unknown setting name
     */
    public function merge(array $overrides): self
    {
        foreach (array_keys($overrides) as $key) {
            if (!in_array($key, self::SETTINGS, true)) {
                throw new UsageException(sprintf('unknown client setting `%s`', Text::bounded($key)));
            }
        }

        return new self(
            templates: self::asString($overrides, 'templates') ?? $this->templates,
            fontDirs: self::asStrings($overrides, 'fontDirs') ?? $this->fontDirs,
            localeDirs: self::asStrings($overrides, 'localeDirs') ?? $this->localeDirs,
            lang: self::asString($overrides, 'lang') ?? $this->lang,
            binary: self::asString($overrides, 'binary') ?? $this->binary,
            logger: self::asLogger($overrides) ?? $this->logger,
            // OR-ed, not overridden: a declared lockdown cannot be lifted.
            strict: $this->strict || (self::asBool($overrides, 'strict') ?? false),
            providers: self::asProviders($overrides) ?? $this->providers,
            env: self::asBool($overrides, 'env') ?? $this->env,
        );
    }

    /**
     * A wrong TYPE is programmer misuse, not a value to ignore: silently
     * inheriting the default would render with a configuration the caller
     * believes they replaced.
     *
     * @param array<string, mixed> $overrides
     */
    private static function asString(array $overrides, string $key): ?string
    {
        $value = $overrides[$key] ?? null;
        if ($value === null) {
            return null;
        }
        if (!is_string($value)) {
            throw new UsageException(sprintf('`%s` takes a string', $key));
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $overrides
     *
     * @return list<string>|null
     */
    private static function asStrings(array $overrides, string $key): ?array
    {
        $value = $overrides[$key] ?? null;
        if ($value === null) {
            return null;
        }
        if (!is_array($value)) {
            throw new UsageException(sprintf('`%s` takes a list of directories', $key));
        }

        $dirs = [];
        foreach ($value as $dir) {
            if (!is_string($dir)) {
                throw new UsageException(sprintf('`%s` takes a list of directories', $key));
            }
            $dirs[] = $dir;
        }

        return $dirs;
    }

    /**
     * @param array<string, mixed> $overrides
     */
    private static function asBool(array $overrides, string $key): ?bool
    {
        $value = $overrides[$key] ?? null;
        if ($value === null) {
            return null;
        }
        if (!is_bool($value)) {
            throw new UsageException(sprintf('`%s` takes a boolean', $key));
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $overrides
     */
    private static function asLogger(array $overrides): ?object
    {
        $value = $overrides['logger'] ?? null;
        if ($value === null) {
            return null;
        }
        if (!is_object($value)) {
            throw new UsageException('`logger` takes an object with a `debug` method');
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $overrides
     *
     * @return array<string, SigningProvider>|null
     */
    private static function asProviders(array $overrides): ?array
    {
        $value = $overrides['providers'] ?? null;
        if ($value === null) {
            return null;
        }
        if (!is_array($value)) {
            throw new UsageException('`providers` takes a map of name to provider');
        }

        $providers = [];
        foreach ($value as $name => $provider) {
            if (!$provider instanceof SigningProvider) {
                throw new UsageException(
                    'a registered signing provider must be a LocalPem or an ExternalSigner',
                );
            }
            $providers[(string) $name] = $provider;
        }

        return $providers;
    }
}
